[core] Defer marking actor dead until confirmed in graceful exit

codope opened the pull request:
Followup to https://github.com/ray-project/ray/pull/57090#discussion_r2511268488

With this patch, the semantics of marking an actor DEAD changes as follows:

- Force kill: Mark DEAD in `DestroyActor` and cleanup.
- Graceful: Store death cause in `DestroyActor` but defer marking DEAD until confirmed in `OnWorkerDead` and then cleanup in `RestartActor`.
- "unsignalable" actor is one where `actor->LocalRayletAddress()` returns nullopt. This happens when:
  - Actor was never scheduled (still PENDING_CREATION)
  - After GCS restart, actors reload from storage without LocalRayletAddress
  For these, we can't send KillLocalActor RPC, so we must mark DEAD immediately rather than waiting for OnWorkerDead (which would never arrive).

codope committed:
[core] Defer marking actor dead until confirmed in graceful exit

Signed-off-by: Sagar Sumit <sagarsumit09@gmail.com>

edoakes commented:
> Force kill: Mark DEAD in DestroyActor and cleanup.

Shouldn't this also wait for confirmation from the Raylet that the process has actually been killed & exited?

codope committed:
Merge remote-tracking branch 'origin/master' into actor-dead-publish

codope committed:
fix detached actor path and address other comments

Signed-off-by: Sagar Sumit <sagarsumit09@gmail.com>

codope committed:
persist state in gcs for graceful out_of_scope shutdown

Signed-off-by: Sagar Sumit <sagarsumit09@gmail.com>

codope committed:
fix actor reconstruction test failure

Signed-off-by: Sagar Sumit <sagarsumit09@gmail.com>

codope committed:
Merge remote-tracking branch 'origin/master' into actor-dead-publish

codope committed:
fix named actor and owner mapping

Signed-off-by: Sagar Sumit <sagarsumit09@gmail.com>

codope committed:
Revert "fix named actor and owner mapping"

This reverts commit a2745d43ce88bbc5de23a064ced58ebcdcd8c5d1.

Signed-off-by: Sagar Sumit <sagarsumit09@gmail.com>

codope committed:
Merge branch 'master' into actor-dead-publish

codope committed:
Merge branch 'master' into actor-dead-publish

codope committed:
update comments

Signed-off-by: Sagar Sumit <sagarsumit09@gmail.com>

codope committed:
Merge remote-tracking branch 'origin/master' into actor-dead-publish

codope committed:
only defer for created actors in DestroyActor

Signed-off-by: Sagar Sumit <sagarsumit09@gmail.com>

codope committed:
fix

Signed-off-by: Sagar Sumit <sagarsumit09@gmail.com>

israbbani commented on src/ray/gcs/gcs_actor_manager.cc:
```diff
@@ -993,6 +1019,12 @@ void GcsActorManager::DestroyActor(const ActorID &actor_id,
+    auto destroyed_it = destroyed_actors_.find(actor_id);
+    if (destroyed_it != destroyed_actors_.end()) {
+      RemoveActorNameFromRegistry(destroyed_it->second);
+    }
```
Are you fixing a separate bug here?

codope replied:
If `DestroyActor` is called multiple times from different paths, the actor might already be in `destroyed_actors_`. We still want to clean up the name. Not a separate bug fix, just being defensive.

israbbani commented on src/ray/gcs/gcs_actor_manager.cc:
```diff
@@ -1082,41 +1120,56 @@ void GcsActorManager::DestroyActor(const ActorID &actor_id,
+  // OnWorkerDead looks up actors in created_actors_ or the scheduler. For actors
+  // in creation phase, CancelActorInScheduling removes them from the scheduler,
+  // so OnWorkerDead won't find them and won't call RestartActor. Also treat
+  // unsignalable actors as needing immediate marking to avoid indefinite waiting.
```
What is an unsignalable actor?

codope replied:
actor without LocalRayletAddress

israbbani commented on src/ray/gcs/gcs_actor_manager.cc:
```diff
@@ -1082,41 +1120,56 @@ void GcsActorManager::DestroyActor(const ActorID &actor_id,
+      AddDestroyedActorToCache(it->second);
+      registered_actors_.erase(it);
+      function_manager_.RemoveJobReference(actor_id.JobId());
+      // Force kills always remove the actor name to allow immediate reuse.
```
This code is executed if the actor is pending creation or missing the local raylet address too (along with force kill).

israbbani commented on src/ray/gcs/gcs_actor_manager.cc:
```diff
@@ -1082,41 +1120,56 @@ void GcsActorManager::DestroyActor(const ActorID &actor_id,
+      RemoveActorNameFromRegistry(actor);
+      if (!actor->IsDetached()) {
+        // For intentional terminations, also remove the owner mapping immediately.
+        if (IsIntentionalTermination(death_cause)) {
```
Why is this check needed now where it wasn't needed before? I think I'm missing something.

israbbani reviewed (commented):
I'm sorry this has taken a while. There's a lot of tech-debt in this area and I'm not convinced I can follow all of the logic. I have a couple of high level questions

1. Do we need to implement the graceful -> forced termination logic inside KillActor as well as DestroyActor? It looks like KillActor is only called when `ray.kill(..., no_restart=False)`

2. While I think setting the node to DEAD before force-killing the worker is a good optimization, I think the code is so hard to read and messy right now that it's simpler/cleaner to keep the behavior consistent. Send the RPC to kill the worker and then mark it as dead after the response is back for all cases. 

3. I didn't follow the logic about unsignalable actors, could you add more context about those to the PR description? 

4. Why did you choose to handle the death notification propagation in the OnWorkerDead handler instead of the response handler to for the KillLocalActor RPC?

codope committed:
change cleanup condition in RestartActor to match IsActorRestartable

Signed-off-by: Sagar Sumit <sagarsumit09@gmail.com>

codope committed:
move ActorTaskSpec deletion inside the cleanup block

Signed-off-by: Sagar Sumit <sagarsumit09@gmail.com>

codope commented on src/ray/gcs/gcs_actor_manager.cc:
```diff
@@ -133,6 +133,32 @@ const ray::rpc::ActorDeathCause GenActorRefDeletedCause(const ray::gcs::GcsActor
 
+// Returns true if the death cause represents an explicit termination due to
+// reference deletion or out-of-scope. These should never restart even if the
+// actor has max_restarts set. RAY_KILL is excluded because it can be used with
```
Correct. `IsExplicitTermination` blocks automatic restarts. Lineage reconstruction uses `HandleRestartActorForLineageReconstruction` with `need_reschedule=true`, which bypasses `should_block_restart`.

codope commented:
> 1. Do we need to implement the graceful -> forced termination logic inside KillActor as well as DestroyActor? It looks like KillActor is only called when `ray.kill(..., no_restart=False)`

KillActor is only called for `ray.kill(..., no_restart=False)` where restarts are allowed. It sends the kill signal and then calls `RestartActor(need_reschedule=true)` to reschedule. `DestroyActor` is for permanent termination (OUT_OF_SCOPE, REF_DELETED, or ray.kill(no_restart=True)). They have slightly different semantics - KillActor allows restart, DestroyActor prevents it - and they can be merged (already there's a followup for that). The graceful shutdown logic only makes sense in DestroyActor.

> 2. While I think setting the node to DEAD before force-killing the worker is a good optimization, I think the code is so hard to read and messy right now that it's simpler/cleaner to keep the behavior consistent. Send the RPC to kill the worker and then mark it as dead after the response is back for all cases.

Valid concern and that's why I have added a lot of comments. The tradeoff was not just optimization but also user experience: consistent behavior (always wait for confirmation) adds latency for force kills. For `ray.kill(force=True)`, users generally expect immediate effect. Open to simplifying if we prefer consistency.

> 3. I didn't follow the logic about unsignalable actors, could you add more context about those to the PR description?

Updated the PR description. An "unsignalable" actor is one where `actor->LocalRayletAddress()` returns nullopt. This happens when:
- Actor was never scheduled (still PENDING_CREATION)
- After GCS restart, actors reload from storage without LocalRayletAddress

For these, we can't send KillLocalActor RPC, so we must mark DEAD immediately rather than waiting for OnWorkerDead (which would never arrive).

> 4. Why did you choose to handle the death notification propagation in the OnWorkerDead handler instead of the response handler to for the KillLocalActor RPC?

OnWorkerDead is the unified path for ALL worker deaths (crash, kill, graceful exit, heartbeat timeout). Using KillLocalActor response would require duplicating cleanup logic. OnWorkerDead is also more reliable - it will eventually fire via heartbeat timeout even if the kill RPC fails or the raylet is down.

codope committed:
fix

Signed-off-by: Sagar Sumit <sagarsumit09@gmail.com>

codope committed:
lint

Signed-off-by: Sagar Sumit <sagarsumit09@gmail.com>

codope committed:
suppress the exception to check if process still alive

Signed-off-by: Sagar Sumit <sagarsumit09@gmail.com>

codope committed:
revert prev and try another fix

Signed-off-by: Sagar Sumit <sagarsumit09@gmail.com>

codope committed:
cleanup

Signed-off-by: Sagar Sumit <sagarsumit09@gmail.com>

codope committed:
Merge branch 'master' into actor-dead-publish

codope committed:
clean refactor gcs_actor_manager; breakup RestartActor, eliminate need_reschedule

Signed-off-by: Sagar Sumit <sagarsumit09@gmail.com>
